File: responsibility/dependency-inversion/payment_service.dart
class PaymentService {

  PaymentService();

  void isPaymentValid() {}
  void openDatabase() {}
  void addNewPayment() {}
  void removePaymentByID() {}
  void updatePaymentByID() {}
}

abstract class Database {
  void insert();
  void update();
  void delete();
}

class MySQLDatabase extends Database {
  @override
  void delete() {
    // implementation
  }

  @override
  void insert() {
    // implementation
  }

  @override
  void update() {
    // implementation
  }
}

class MongoDatabase extends Database {
  @override
  void delete() {
    // implementation
  }

  @override
  void insert() {
    // implementation
  }

  @override
  void update() {
    // implementation
  }
}

File: responsibility/dependency-inversion/PaymentService.kt
class PaymentService (val database: Database) {

  fun paymentIsValid() {
     // Implementation code
  }

  fun openDatabase() {
     // Implementation code
  }

  fun addNewPayment() {
     // Implementation code
  }

  fun removePaymentByID() {
     // Implementation code
  }

  fun updatePaymentByID() {
     // Implementation code
  }
}

abstract class Database {
  abstract fun insert()
  abstract fun update()
  abstract fun delete()
}

class MySQLDatabase : Database() {

  override fun insert() {
      // Implementation code
  }

  override fun update() {
      // Implementation code
  }

  override fun delete() {
      // Implementation code
  }
}

class MongoDatabase : Database() {
  override fun insert() {
      // Implementation code
  }

  override fun update() {
      // Implementation code
  }

  override fun delete() {
      // Implementation code
  }
}
